tutorial/02_image_filtering.md

```diff
@@ -0,0 +1,821 @@
+<!-- #endregion -->
+
+<!-- #region editable=true slideshow={"slide_type": ""} -->
+Incidentally, for reference, let's have a look at what the Gaussian filter actually looks like. Technically, the value of the kernel at a pixel that is $r$ rows and $c$ cols from the center is:
+
+$$
+k_{r, c} = \frac{1}{2\pi \sigma^2} \exp{\left(-\frac{r^2 + c^2}{2\sigma^2}\right)}
+$$
+
+Practically speaking, this value is pretty close to zero for values more than $4\sigma$ away from the center, so practical Gaussian filters are truncated at about $4\sigma$:
+<!-- #endregion -->
+
+```python
+sidelen = 45
+sigma = (sidelen - 1) // 2 // 4
+spot = np.zeros((sidelen, sidelen), dtype=float)
+spot[sidelen // 2, sidelen // 2] = 1
+kernel = ski.filters.gaussian(spot, sigma=sigma)
+
+imshow_all(spot, kernel / np.max(kernel))
+```
+
+<!-- #region editable=true slideshow={"slide_type": ""} -->
+If we have a look at a slice of the filtered result at its midpoint, we see the classic shape of a Gaussian kernel in one dimension:
+<!-- #endregion -->
+
+```python
+fig, ax = plt.subplots(1, 2, figsize=[9, 3], width_ratios=[1, 1.5])
+
+ax[0].imshow(kernel, cmap='gray')
+ax[0].vlines(kernel.shape[1] // 2, -100, 100)
+ax[0].set_ylim((sidelen - 1, 0))
+
+ax[1].plot(kernel[:, kernel.shape[1] // 2])
+```
+
+<!-- #region editable=true slideshow={"slide_type": "slide"} -->
+## Basic edge filtering
+<!-- #endregion -->
+
+<!-- #region editable=true slideshow={"slide_type": "notes"} -->
+For images, edges are boundaries between light and dark values. The detection of edges can be useful on its own, or it can be used as preliminary step in other algorithms (which we'll see later).
+<!-- #endregion -->
+
+<!-- #region editable=true slideshow={"slide_type": "slide"} -->
+### Difference filters in 2D
+<!-- #endregion -->
+
+<!-- #region slideshow={"slide_type": "notes"} -->
+For images, you can think of an edge as points where the gradient is large in one direction. We can approximate gradients with difference filters.
+<!-- #endregion -->
+
+```python slideshow={"slide_type": "fragment"}
+vertical_kernel = np.array([
+    [-1],
+    [ 0],
+    [ 1],
+])
+
+gradient_vertical = ndi.correlate(pixelated.astype(float),
+                                  vertical_kernel)
+fig, ax = plt.subplots()
+ax.imshow(gradient_vertical, cmap="gray");
+```
+
+```python
+vertical_kernel.shape
+```
+
+<!-- #region slideshow={"slide_type": "fragment"} -->
+### Exercise 3:
+<!-- #endregion -->
+
+- Add a horizontal kernel to the above example to also compute the horizontal gradient, $g_y$
+- Compute the magnitude of the image gradient at each point: $\left|g\right| = \sqrt{g_x^2 + g_y^2}$
+
+```python
+# Solution here
+```
+
+```python
+# Solution here
+```
+
+<!-- #region editable=true slideshow={"slide_type": "slide"} -->
+### Sobel edge filter
```

Comment:
I've done away completely with the machine learning based exercise. Still thinking about what else to trim. I'm aiming for 30-40 minutes for this tutorial depending how much time is needed by the 3rd part.